実践課題用ドキュメントの作成 (#8)
jaws-ug-cdk/cdk-conf-2024-contribute-workshop: src/content/docs/5-実践課題/tag-mutability.md

   * If IMMUTABLE is specified, all image tags within the repository will be immutable which will prevent them from being overwritten.
   *
   *  @default ImageTagMutability.MUTABLE
   */
  readonly imageTagMutability?: ImageTagMutability;
};
```

#### `Repository`クラスのconstructor関数の変更

`Repository`クラスのconstructor関数内で、`CfnRepository`クラスのインスタンス作成時に`imageTagMutability`引数を設定します。

```ts {8}
// aws-cdk/packages/aws-cdk-lib/aws-ecr/lib/repository.ts
export class Repository extends RepositoryBase {
  constructor(scope: Construct, id: string, props: RepositoryProps) {
    super(scope, id, props);

    const resource = new CfnRepository(this, 'Resource', {
      ..., // 既存のプロパティ
      imageTagMutability: props.imageTagMutability
    });
  };
};
```

以上で機能修正は完了です。一連の内容をcommitしておきましょう。

### ユニットテストの追加

#### テストファイルの修正

続いてユニットテストを追加します。テストファイルは`aws-cdk/packages/aws-cdk-lib/aws-ecr/test/repository.test.ts`です。

ユニットテストでは、`imageTagMutability`プロパティがCloudformationテンプレートに正しく反映されるかを確認します。

```ts {2-12}
// aws-cdk/packages/aws-cdk-lib/aws-ecr/test/repository.test.ts
test('specify imageTagMutability', () => {
  const stack = new cdk.Stack();

>   new ecr.repository(stack, 'Myrepository', {

Reviewer: ```suggestion
  new ecr.repository(stack, 'MyRepository', {
```

     imageTagMutability: ecr.ImageTagMutability.IMMUTABLE,
  });

>   Template.fromStack(stack).hasResourceProperties('AWS::ecr::Repository', {

Reviewer: ```suggestion
  Template.fromStack(stack).hasResourceProperties('AWS::ECR::Repository', {
```

// 統合テストの実行
new integ.IntegTest(app, 'EcrTest', {
  testCases: [stack],
});
```

#### 統合テストの実行

統合テストを実行して、テストが正常に通ることを確認します。

```sh
# repositoryクラスを含めたecrのコンストラクトを再ビルド
cd /packages/aws-cdk-lib/aws-ecr
yarn tsc

cd packages/@aws-cdk-testing/framework-integ/test/aws-ecr
# integ ファイルのビルド/トランスパイルをして、javascript ファイルを生成
yarn tsc
# 実際にintegテストを実行する
yarn integ aws-ecr/test/integ.ecr-image-tag-immutability.js --update-on-failed
```

自動でCloudformationテンプレートの作成, スタックのデプロイ及び削除が行われます。エラーなく完了すれば統合テストも完了です。

作成したテストファイル及び生成されたスナップショットファイルをコミットしておきましょう。

### READMEの更新

最後に、この機能追加に関する説明をREADMEに追加します。

```md
// aws-cdk/packages/aws-cdk-lib/aws-ecr/README.md
### Image tag immutability

You can set tag immutability on images in our repository using the `imageTagMutability` construct prop.

```ts
new ecr.Repository(this, 'Repo', {
  imageTagMutability: ecr.ImageTagMutability.IMMUTABLE,
});
\```
```

以上で必要な変更はすべて完了しました。

### PRの提出

最後にPRを提出します。

[PRルール](/cdk-conf-2024-contribute-workshop/2-コントリビュートの流れとルール/contribution-flow-rule/#pr-ルール)に則り、以下のようなPRを作成しましょう。

|項目|内容|記入例|
|-|-|-|
|PRタイトル|PRの目的|feat(ecr): add `imageTagMutability` property to a repository|
|Issue # (if applicable)|関連するIssue番号|Closes #2|

> |Reason for this change|変更理由|We can set a image tag immuutability for an ecr repository from cloudformation, but this was not supported in the AWS CDK L2 construct.|

Reviewer: ```suggestion
|Reason for this change|変更理由|We can set an image tag immutability for an ecr repository from cloudformation, but this was not supported in the AWS CDK L2 construct.|
```
